context.go

```diff
@@ -0,0 +1,196 @@
+package inject
+
+import (
+	"reflect"
+	"sort"
+	"strings"
+)
+
+// ctx is the dependency resolution context. It is used to detect circular
+// dependencies and provide the dependency tree.
+type ctx struct {
+	root    *stack
+	current *stack
+}
+
+func zeroCtx() ctx {
```

Comment:
Looks like this isn't used anywhere? Also, just curious, what is the reason for making it a function, rather than just a variable?

Reply:
Right, not used anymore in final version. Will remove. 
A function guarantees that nobody can change the value, which is not the case with global variables. Also, the function will be inlined by the compiler, so is equivalent to actually creating a zero value in the calling code (i.e. `ctx{}`).